design/user-event-spi.md

```diff
@@ -0,0 +1,149 @@
+# User Event SPI
+
+* **Status**: Draft v0.1
+* **Author**: [vmuzikar](https://github.com/vmuzikar)
+
+## Motivation
+
+Current Event SPI design has several limitations. E.g. it's not defined when an event is fired exactly, there are no pre-
+and post-events. There's also no way of pre-events intercepting for approve/reject functionality.
+
+## Use cases
+
+* A user is removed from Keycloak. All related records in an external system are removed before the user is permanently
+  deleted (e.g. eshop orders).
+* A user self-registers in Keycloak. Their registration must be approved by an external system (either human or automated)
+  before the new account is actually fully created.
+* A user self-edits their email address. The change must be approved first by an external system before it's actually
+  applied.
+
+## Design proposal
```

Comment:
There's one thing missing here in the Design Proposal. The Event dispatcher should be run on a separate and independent thread so it doesn't block Keycloak Operations